### ISSUE TITLE
Implement bt* (bit test) instructions in validator

### ISSUE THREAD

**Issue by jrkoenig** [StanfordPL/stoke#934]
Add support for bt, btc, btr, bts. These instructions cannot use the simple handler because the memory forms require adjusting the address in the memory `Operand` before actually accessing memory.

**Comment by bchurchill** [comment]
SimpleHandler should still be able to work.  It gives you a reference to the SymState (typically called ss), and you can use `ss->memory` to access the full interface of the memory model abstraction.  You'll be able to use most helper functions as usual, just not some of the helper functions that use the memory model for you.

**Comment by jrkoenig** [comment]
You're right, I put my stuff in simple_handler. I've got almost everything working, but there's issues caused by incorrectness in the sandbox. The bt instructions don't access memory normally, so the sandbox rewrites them internally (see `src/sandbox/sandbox.cc:1145`). This rewrite changes their behavior. According to the intel manual, 8bit constant operands are masked according to the defined width of the instruction. So a `btcw 0xFF, (%rax)` is effectively `btcw 0x0F, (%rax)`, and a `btcl 0xFF, (%rax)` is effectively `btcl 0x1F, (%rax)`, etc. Also, if a register is supplied, it is sign extended, not zero extended. Instructions like `btcw %dx, (%rax)` get translated to `%rsi = 0; movw %dx, %si; btcq %rsi, (%rsi)`. This makes the writes incorrect because the move is implicitly zero-extending.

Testing raw instructions with GDB matches what the manual says. I'm 99% sure my circuits also match the manual, as I was very careful about sign versus zero extension, but the fuzz tester said the circuits did not match the sandbox which lead me to investigate. Note this only effects memory read or write operations, as the register variants are not transliterated in this fashion.

**Comment by bchurchill** [comment]
Good find!  It's indeed fairly common for us to discover sandbox/assembler bugs when we implement validator circuits for the same time.  Rarely we've also discovered situations where both the sandbox and validator were wrong but agreed with each other; this typically only happens with instructions that need special sandbox support.  For the longest time our push/pop were broken, but the operations were symmetric and the sandbox/validator agreed on what they were.  (sorry I commented twice earlier, I think I had written the email but it didn't show up on the site, and then later I just assumed that I hadn't sent it... but then both showed up)
